In [3]:
import pandas as pd

fake_df = pd.read_csv(r"Dataset/DataSet_Misinfo_FAKE.csv")
true_df = pd.read_csv(r"Dataset/DataSet_Misinfo_TRUE.csv")
propaganda_df = pd.read_csv(r"Dataset/EXTRA_RussianPropagandaSubset.csv")



In [4]:
import pandas as pd
import string
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer, WordNetLemmatizer
import nltk
%matplotlib inline
import matplotlib
matplotlib.use('Agg')


In [5]:
# Summary statistics for fake_df
print("Summary statistics for DataSet_Misinfo_FAKE.csv:")
print(fake_df.describe())

Summary statistics for DataSet_Misinfo_FAKE.csv:
         Unnamed: 0
count  43642.000000
mean   22293.806173
std    12889.800176
min        0.000000
25%    10910.250000
50%    22450.500000
75%    33472.750000
max    44426.000000


In [6]:
# Summary statistics for true_df
print("\nSummary statistics for DataSet_Misinfo_TRUE.csv:")
print(true_df.describe())


Summary statistics for DataSet_Misinfo_TRUE.csv:
         Unnamed: 0
count  34975.000000
mean   17487.000000
std    10096.557169
min        0.000000
25%     8743.500000
50%    17487.000000
75%    26230.500000
max    34974.000000


In [7]:
# Summary statistics for propaganda_df
print("\nSummary statistics for EXTRA_RussianPropagandaSubset.csv:")
print(propaganda_df.describe())



Summary statistics for EXTRA_RussianPropagandaSubset.csv:
       Unnamed: 0
count   7369.0000
mean    3684.0000
std     2127.3914
min        0.0000
25%     1842.0000
50%     3684.0000
75%     5526.0000
max     7368.0000


In [8]:
# Adding a new column "Label" with "true" for true_df and "fake" for fake_df
true_df["Label"] = "true"
fake_df["Label"] = "fake"

In [9]:
# Concatenating the two dataframes into a single dataframe
combined_df = pd.concat([true_df, fake_df], ignore_index=True)

# Print the first few rows of the combined dataframe to verify the new column has been added
print(combined_df.head())

   Unnamed: 0                                               text Label
0           0  The head of a conservative Republican faction ...  true
1           1  Transgender people will be allowed for the fir...  true
2           2  The special counsel investigation of links bet...  true
3           3  Trump campaign adviser George Papadopoulos tol...  true
4           4  President Donald Trump called on the U.S. Post...  true


In [10]:
# Removing the index column from the combined_df dataframe
combined_df = combined_df.drop(columns=["Unnamed: 0"])

# Print the first few rows of the updated combined dataframe to verify the index column has been removed
print(combined_df.head())

                                                text Label
0  The head of a conservative Republican faction ...  true
1  Transgender people will be allowed for the fir...  true
2  The special counsel investigation of links bet...  true
3  Trump campaign adviser George Papadopoulos tol...  true
4  President Donald Trump called on the U.S. Post...  true


In [11]:
combined_df['Label'] = combined_df['Label'].replace({'true': True, 'fake': False})


In [12]:
combined_df['text'] = combined_df['text'].astype(str)

In [13]:
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.stem import PorterStemmer, WordNetLemmatizer
from nltk.corpus import wordnet
import re

# Add a new column to store the count of stop words in every row of the dataset in the 'text' column
stop_words = set(stopwords.words('english'))
combined_df['stop_words_count'] = combined_df['text'].apply(lambda x: len([word for word in word_tokenize(x) if word.lower() in stop_words]))

# # Stem the words in the 'text' column
# stemmer = PorterStemmer()
# combined_df['stemmed_text'] = combined_df['text'].apply(lambda x: ' '.join(stemmer.stem(word) for word in word_tokenize(x)))

# Map part of speech tags to WordNet tags
# def get_wordnet_pos(word):
#     """Map POS tag to first character used by WordNetLemmatizer"""
#     tag = nltk.pos_tag([word])[0][1][0].upper()
#     tag_dict = {"J": wordnet.ADJ,
#                 "N": wordnet.NOUN,
#                 "V": wordnet.VERB,
#                 "R": wordnet.ADV}
#     return tag_dict.get(tag, wordnet.NOUN)

# # Lemmatize the words in the 'text' column
# lemmatizer = WordNetLemmatizer()
# combined_df['lemmatized_text'] = combined_df['text'].apply(lambda x: ' '.join(lemmatizer.lemmatize(word, get_wordnet_pos(word)) for word in word_tokenize(x)))


In [14]:
# Word count
combined_df['word_count'] = combined_df['text'].apply(lambda x: len(str(x).split(" ")))

# Capitalization count
combined_df['cap_count'] = combined_df['text'].apply(lambda x: sum(1 for c in x if c.isupper()))

# Punctuation count
combined_df['punct_count'] = combined_df['text'].apply(lambda x: len([c for c in x if c in string.punctuation]))

# Number of exclamation marks in headline or summary
combined_df['num_exclamations'] = combined_df['text'].apply(lambda x: x.count('!')) + combined_df['text'].apply(lambda x: x.count('!'))

# Number of capitalized words in headline or summary
combined_df['cap_words_headline'] = combined_df['text'].apply(lambda x: sum(1 for c in x if c.isupper())) 
combined_df['cap_words_summary'] = combined_df['text'].apply(lambda x: sum(1 for c in x if c.isupper()))



In [15]:
combined_df

text  Label  \
0      The head of a conservative Republican faction ...   True   
1      Transgender people will be allowed for the fir...   True   
2      The special counsel investigation of links bet...   True   
3      Trump campaign adviser George Papadopoulos tol...   True   
4      President Donald Trump called on the U.S. Post...   True   
...                                                  ...    ...   
78612  The USA wants to divide Syria.\r\n\r\nGreat Br...  False   
78613  The Ukrainian coup d'etat cost the US nothing ...  False   
78614  The European Parliament falsifies history by d...  False   
78615  The European Parliament falsifies history by d...  False   
78616  A leading FSB officer, Segey Beseda, said duri...  False   

       stop_words_count  word_count  cap_count  punct_count  num_exclamations  \
0                   303         749        144          115                 0   
1                   171         396         72           40                 0   
2                   201         462        106           44                 0   
3                   153         373         99           48                 0   
4                   348         860        182          132                 2   
...                 ...         ...        ...          ...               ...   
78612                36          70         26           14                 0   
78613                40         106         38           13                 0   
78614                61         129         28           16                 0   
78615                61         129         28           16                 0   
78616                24          55         26            6                 0   

       cap_words_headline  cap_words_summary  
0                     144                144  
1                      72                 72  
2                     106                106  
3                      99                 99  
4                     182                182  
...                   ...                ...  
78612                  26                 26  
78613                  38                 38  
78614                  28                 28  
78615                  28                 28  
78616                  26                 26  

[78617 rows x 9 columns]

In [16]:
# Initial comparision of labels and stop words

import seaborn as sns
import matplotlib.pyplot as plt
# Create a violinplot to visualize the distribution of stop words count with respect to the label column
sns.violinplot(x='Label', y='stop_words_count', data=combined_df)

# Set the x-label and y-label
plt.xlabel('Label')
plt.ylabel('Stop Words Count')

# Set the plot title
plt.title('Stop Words Count vs Label')

# Save the plot as a PNG file
plt.savefig('stop_words_count_vs_label.png')


In [17]:
# let's look deeper into the distribution of the previous plot


import seaborn as sns
import matplotlib.pyplot as plt

# Create a scatter plot to visualize the relationship between word count and stop words count with respect to the label column
sns.scatterplot(x='word_count', y='stop_words_count', hue='Label', data=combined_df)

# Set the x-label and y-label
plt.xlabel('Word Count')
plt.ylabel('Stop Words Count')

# Set the plot title
plt.title('Word Count vs Stop Words Count, by Label')

# Save the plot as a PNG file
plt.savefig('word_count_vs_stop_words_count.png')




In [18]:
# Calculate the total number of words and stop words in each text
combined_df['total_words'] = combined_df['word_count'] + combined_df['num_exclamations']
combined_df['total_stop_words'] = combined_df['stop_words_count'] + combined_df['num_exclamations']

# Calculate the ratio of stop words to total words for each text
combined_df['stop_words_ratio'] = combined_df['total_stop_words'] / combined_df['total_words']

# Calculate the mean stop words ratio for true and fake news
true_mean_ratio = combined_df[combined_df['Label']==True]['stop_words_ratio'].mean()
fake_mean_ratio = combined_df[combined_df['Label']==False]['stop_words_ratio'].mean()

print(f"True news stop words ratio: {true_mean_ratio:.2%}")
print(f"Fake news stop words ratio: {fake_mean_ratio:.2%}")




True news stop words ratio: 40.74%
Fake news stop words ratio: 42.58%


This indicates that the number of stop words used in the text may not be a very strong indicator of whether a news article is true or fake. </br>
Other factors may have a greater impact on determining the veracity of a news article.

In [19]:
# Calculate stop words ratio for true news
true_df = combined_df[combined_df['Label'] == True]
true_stop_words_count = true_df['stop_words_count'].sum()
true_word_count = true_df['word_count'].sum()
true_stop_words_ratio = true_stop_words_count / true_word_count
print(f"True news stop words ratio: {true_stop_words_ratio:.2%}")

# Calculate stop words ratio for fake news
fake_df = combined_df[combined_df['Label'] == False]
fake_stop_words_count = fake_df['stop_words_count'].sum()
fake_word_count = fake_df['word_count'].sum()
fake_stop_words_ratio = fake_stop_words_count / fake_word_count
print(f"Fake news stop words ratio: {fake_stop_words_ratio:.2%}")


True news stop words ratio: 42.23%
Fake news stop words ratio: 43.43%


In [20]:
# Group the data by label and compute the average word count for each group
avg_word_count = combined_df.groupby('Label')['word_count'].mean()

# Print the results
print(avg_word_count)

Label
False    442.507080
True     543.208006
Name: word_count, dtype: float64


looks like false news in this dataset has fewer average words compared to true news. 

In [21]:
# Calculate the normalized cap count by dividing it with the word count
normalized_cap_count = combined_df['cap_count'] / combined_df['word_count']

# Create a boxplot to visualize the distribution of normalized cap count with respect to the label column
sns.boxplot(x='Label', y=normalized_cap_count, data=combined_df)

# Set the x-label and y-label
plt.xlabel('Label')
plt.ylabel('Normalized Cap Count')

# Set the plot title
plt.title('Normalized Cap Count vs Label')

# Save the plot as a PNG file
plt.savefig('normalized_cap_count_vs_label.png')

# Print the mean normalized cap count for true and fake news
print("Mean normalized cap count for true news: {:.2%}".format(normalized_cap_count[combined_df['Label'] == True].mean()))
print("Mean normalized cap count for fake news: {:.2%}".format(normalized_cap_count[combined_df['Label'] == False].mean()))


Mean normalized cap count for true news: 20.79%
Mean normalized cap count for fake news: 28.63%


capital letters are more in fake news

In [22]:
# Calculate normalized punctuation count for each row
combined_df['punct_count_norm'] = combined_df['punct_count'] / combined_df['word_count']

# Calculate mean normalized punctuation count for true and fake news
true_punct_mean = combined_df[combined_df['Label']==True]['punct_count_norm'].mean()
fake_punct_mean = combined_df[combined_df['Label']==False]['punct_count_norm'].mean()

print(f"Mean normalized punctuation count for true news: {true_punct_mean:.2%}")
print(f"Mean normalized punctuation count for fake news: {fake_punct_mean:.2%}")


Mean normalized punctuation count for true news: 12.57%
Mean normalized punctuation count for fake news: 16.38%


punctuations are more in fake news. 

In [23]:
# Calculate the normalized number of exclamation marks for each row
combined_df['norm_num_exclamations'] = combined_df['num_exclamations'] / combined_df['word_count']

# Calculate the mean normalized number of exclamation marks for each label
mean_norm_excl_true = combined_df[combined_df['Label']==True]['norm_num_exclamations'].mean()
mean_norm_excl_fake = combined_df[combined_df['Label']==False]['norm_num_exclamations'].mean()

# Print the results
print("Mean normalized number of exclamation marks for true news: {:.2%}".format(mean_norm_excl_true))
print("Mean normalized number of exclamation marks for fake news: {:.2%}".format(mean_norm_excl_fake))


Mean normalized number of exclamation marks for true news: 0.06%
Mean normalized number of exclamation marks for fake news: 0.59%


In [24]:
combined_df = combined_df.drop('cap_words_headline', axis=1)

In [25]:
combined_df

text  Label  \
0      The head of a conservative Republican faction ...   True   
1      Transgender people will be allowed for the fir...   True   
2      The special counsel investigation of links bet...   True   
3      Trump campaign adviser George Papadopoulos tol...   True   
4      President Donald Trump called on the U.S. Post...   True   
...                                                  ...    ...   
78612  The USA wants to divide Syria.\r\n\r\nGreat Br...  False   
78613  The Ukrainian coup d'etat cost the US nothing ...  False   
78614  The European Parliament falsifies history by d...  False   
78615  The European Parliament falsifies history by d...  False   
78616  A leading FSB officer, Segey Beseda, said duri...  False   

       stop_words_count  word_count  cap_count  punct_count  num_exclamations  \
0                   303         749        144          115                 0   
1                   171         396         72           40                 0   
2                   201         462        106           44                 0   
3                   153         373         99           48                 0   
4                   348         860        182          132                 2   
...                 ...         ...        ...          ...               ...   
78612                36          70         26           14                 0   
78613                40         106         38           13                 0   
78614                61         129         28           16                 0   
78615                61         129         28           16                 0   
78616                24          55         26            6                 0   

       cap_words_summary  total_words  total_stop_words  stop_words_ratio  \
0                    144          749               303          0.404539   
1                     72          396               171          0.431818   
2                    106          462               201          0.435065   
3                     99          373               153          0.410188   
4                    182          862               350          0.406032   
...                  ...          ...               ...               ...   
78612                 26           70                36          0.514286   
78613                 38          106                40          0.377358   
78614                 28          129                61          0.472868   
78615                 28          129                61          0.472868   
78616                 26           55                24          0.436364   

       punct_count_norm  norm_num_exclamations  
0              0.153538               0.000000  
1              0.101010               0.000000  
2              0.095238               0.000000  
3              0.128686               0.000000  
4              0.153488               0.002326  
...                 ...                    ...  
78612          0.200000               0.000000  
78613          0.122642               0.000000  
78614          0.124031               0.000000  
78615          0.124031               0.000000  
78616          0.109091               0.000000  

[78617 rows x 13 columns]

In [26]:
pip install vaderSentiment

Note: you may need to restart the kernel to use updated packages.


use the VADER (Valence Aware Dictionary and Sentiment Reasoner) sentiment analysis tool to analyze the sentiment of the text in the text column.

First, we need to install the vaderSentiment library which contains the VADER sentiment analysis tool. We can do this by running !pip install vaderSentiment in a code cell.

Then we can import the SentimentIntensityAnalyzer class from the vaderSentiment library and create an instance of the class to use for sentiment analysis.

After that, we can define a function that takes a piece of text as input, uses the SentimentIntensityAnalyzer to analyze the sentiment of the text, and returns the sentiment scores as a dictionary.

Finally, we can apply this function to the text column of the dataframe using the apply() method and create a new column in the dataframe to store the sentiment scores.

In [27]:
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer

# Create an instance of the SentimentIntensityAnalyzer class
analyzer = SentimentIntensityAnalyzer()

# Define a function to analyze the sentiment of a piece of text
def analyze_sentiment(text):
    scores = analyzer.polarity_scores(text)
    return scores

# Apply the analyze_sentiment function to the text column of the dataframe and create a new column to store the sentiment scores
combined_df['sentiment_scores'] = combined_df['text'].apply(analyze_sentiment)

In [28]:
combined_df

text  Label  \
0      The head of a conservative Republican faction ...   True   
1      Transgender people will be allowed for the fir...   True   
2      The special counsel investigation of links bet...   True   
3      Trump campaign adviser George Papadopoulos tol...   True   
4      President Donald Trump called on the U.S. Post...   True   
...                                                  ...    ...   
78612  The USA wants to divide Syria.\r\n\r\nGreat Br...  False   
78613  The Ukrainian coup d'etat cost the US nothing ...  False   
78614  The European Parliament falsifies history by d...  False   
78615  The European Parliament falsifies history by d...  False   
78616  A leading FSB officer, Segey Beseda, said duri...  False   

       stop_words_count  word_count  cap_count  punct_count  num_exclamations  \
0                   303         749        144          115                 0   
1                   171         396         72           40                 0   
2                   201         462        106           44                 0   
3                   153         373         99           48                 0   
4                   348         860        182          132                 2   
...                 ...         ...        ...          ...               ...   
78612                36          70         26           14                 0   
78613                40         106         38           13                 0   
78614                61         129         28           16                 0   
78615                61         129         28           16                 0   
78616                24          55         26            6                 0   

       cap_words_summary  total_words  total_stop_words  stop_words_ratio  \
0                    144          749               303          0.404539   
1                     72          396               171          0.431818   
2                    106          462               201          0.435065   
3                     99          373               153          0.410188   
4                    182          862               350          0.406032   
...                  ...          ...               ...               ...   
78612                 26           70                36          0.514286   
78613                 38          106                40          0.377358   
78614                 28          129                61          0.472868   
78615                 28          129                61          0.472868   
78616                 26           55                24          0.436364   

       punct_count_norm  norm_num_exclamations  \
0              0.153538               0.000000   
1              0.101010               0.000000   
2              0.095238               0.000000   
3              0.128686               0.000000   
4              0.153488               0.002326   
...                 ...                    ...   
78612          0.200000               0.000000   
78613          0.122642               0.000000   
78614          0.124031               0.000000   
78615          0.124031               0.000000   
78616          0.109091               0.000000   

                                        sentiment_scores  
0      {'neg': 0.048, 'neu': 0.866, 'pos': 0.086, 'co...  
1      {'neg': 0.063, 'neu': 0.837, 'pos': 0.101, 'co...  
2      {'neg': 0.049, 'neu': 0.907, 'pos': 0.044, 'co...  
3      {'neg': 0.066, 'neu': 0.875, 'pos': 0.059, 'co...  
4      {'neg': 0.043, 'neu': 0.898, 'pos': 0.059, 'co...  
...                                                  ...  
78612  {'neg': 0.0, 'neu': 0.901, 'pos': 0.099, 'comp...  
78613  {'neg': 0.082, 'neu': 0.891, 'pos': 0.026, 'co...  
78614  {'neg': 0.146, 'neu': 0.796, 'pos': 0.058, 'co...  
78615  {'neg': 0.146, 'neu': 0.796, 'pos': 0.058, 'co...  
78616  {'neg': 0.066, 'neu': 0.835, 'pos': 0.1, 'comp...  

[78617 rows x 14 columns]

In [29]:
# Create a new column 'sentiment_scores_num' with numerical sentiment scores
combined_df['sentiment_scores_num'] = combined_df['sentiment_scores'].apply(lambda x: x['compound'])

In [30]:
import matplotlib.pyplot as plt

# Separate the data based on label
true_df = combined_df[combined_df['Label']]
false_df = combined_df[~combined_df['Label']]

# Create a new column 'sentiment_scores_num' with numerical sentiment scores
combined_df['sentiment_scores_num'] = combined_df['sentiment_scores'].apply(lambda x: x['compound'])

# Plot histograms of sentiment scores for each label
plt.hist(true_df['sentiment_scores_num'], bins=20, alpha=0.5, label='True')
plt.hist(false_df['sentiment_scores_num'], bins=20, alpha=0.5, label='False')
plt.legend(loc='upper left')
plt.xlabel('Sentiment Score')
plt.ylabel('Frequency')
plt.title('Distribution of Sentiment Scores by Label')
plt.savefig('sentiment_histogram.png')


In [31]:
combined_df

text  Label  \
0      The head of a conservative Republican faction ...   True   
1      Transgender people will be allowed for the fir...   True   
2      The special counsel investigation of links bet...   True   
3      Trump campaign adviser George Papadopoulos tol...   True   
4      President Donald Trump called on the U.S. Post...   True   
...                                                  ...    ...   
78612  The USA wants to divide Syria.\r\n\r\nGreat Br...  False   
78613  The Ukrainian coup d'etat cost the US nothing ...  False   
78614  The European Parliament falsifies history by d...  False   
78615  The European Parliament falsifies history by d...  False   
78616  A leading FSB officer, Segey Beseda, said duri...  False   

       stop_words_count  word_count  cap_count  punct_count  num_exclamations  \
0                   303         749        144          115                 0   
1                   171         396         72           40                 0   
2                   201         462        106           44                 0   
3                   153         373         99           48                 0   
4                   348         860        182          132                 2   
...                 ...         ...        ...          ...               ...   
78612                36          70         26           14                 0   
78613                40         106         38           13                 0   
78614                61         129         28           16                 0   
78615                61         129         28           16                 0   
78616                24          55         26            6                 0   

       cap_words_summary  total_words  total_stop_words  stop_words_ratio  \
0                    144          749               303          0.404539   
1                     72          396               171          0.431818   
2                    106          462               201          0.435065   
3                     99          373               153          0.410188   
4                    182          862               350          0.406032   
...                  ...          ...               ...               ...   
78612                 26           70                36          0.514286   
78613                 38          106                40          0.377358   
78614                 28          129                61          0.472868   
78615                 28          129                61          0.472868   
78616                 26           55                24          0.436364   

       punct_count_norm  norm_num_exclamations  \
0              0.153538               0.000000   
1              0.101010               0.000000   
2              0.095238               0.000000   
3              0.128686               0.000000   
4              0.153488               0.002326   
...                 ...                    ...   
78612          0.200000               0.000000   
78613          0.122642               0.000000   
78614          0.124031               0.000000   
78615          0.124031               0.000000   
78616          0.109091               0.000000   

                                        sentiment_scores  sentiment_scores_num  
0      {'neg': 0.048, 'neu': 0.866, 'pos': 0.086, 'co...                0.9847  
1      {'neg': 0.063, 'neu': 0.837, 'pos': 0.101, 'co...                0.9496  
2      {'neg': 0.049, 'neu': 0.907, 'pos': 0.044, 'co...               -0.6808  
3      {'neg': 0.066, 'neu': 0.875, 'pos': 0.059, 'co...               -0.2201  
4      {'neg': 0.043, 'neu': 0.898, 'pos': 0.059, 'co...                0.6683  
...                                                  ...                   ...  
78612  {'neg': 0.0, 'neu': 0.901, 'pos': 0.099, 'comp...                0.7650  
78613  {'neg': 0.082, 'neu': 0.891, 'pos': 0.026, 'co...               -0.6492  
78614  {'neg': 0.14

Here this is important. True news is more positive in this dataset


In [32]:
import matplotlib.pyplot as plt

# Separate the data based on label
true_df = combined_df[combined_df['Label']]
false_df = combined_df[~combined_df['Label']]

# Plot box plots of sentiment scores for each label
fig, ax = plt.subplots()
ax.boxplot([true_df['sentiment_scores_num'], false_df['sentiment_scores_num']])
ax.set_xticklabels(['True', 'False'])
ax.set_ylabel('Sentiment Score')
ax.set_title('Distribution of Sentiment Scores by Label')
plt.savefig('sentiment_boxplot.png')


In [33]:
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.stem import WordNetLemmatizer

stop_words = set(stopwords.words('english'))
lemmatizer = WordNetLemmatizer()

def preprocess_text(text):
    # Tokenize the text
    tokens = word_tokenize(text.lower())

    # Remove stopwords and non-alphabetic characters
    tokens = [token for token in tokens if token.isalpha() and token not in stop_words]

    # Lemmatize the tokens
    tokens = [lemmatizer.lemmatize(token) for token in tokens]

    return tokens

In [35]:
combined_df

text  Label  \
0      The head of a conservative Republican faction ...   True   
1      Transgender people will be allowed for the fir...   True   
2      The special counsel investigation of links bet...   True   
3      Trump campaign adviser George Papadopoulos tol...   True   
4      President Donald Trump called on the U.S. Post...   True   
...                                                  ...    ...   
78612  The USA wants to divide Syria.\r\n\r\nGreat Br...  False   
78613  The Ukrainian coup d'etat cost the US nothing ...  False   
78614  The European Parliament falsifies history by d...  False   
78615  The European Parliament falsifies history by d...  False   
78616  A leading FSB officer, Segey Beseda, said duri...  False   

       stop_words_count  word_count  cap_count  punct_count  num_exclamations  \
0                   303         749        144          115                 0   
1                   171         396         72           40                 0   
2                   201         462        106           44                 0   
3                   153         373         99           48                 0   
4                   348         860        182          132                 2   
...                 ...         ...        ...          ...               ...   
78612                36          70         26           14                 0   
78613                40         106         38           13                 0   
78614                61         129         28           16                 0   
78615                61         129         28           16                 0   
78616                24          55         26            6                 0   

       cap_words_summary  total_words  total_stop_words  stop_words_ratio  \
0                    144          749               303          0.404539   
1                     72          396               171          0.431818   
2                    106          462               201          0.435065   
3                     99          373               153          0.410188   
4                    182          862               350          0.406032   
...                  ...          ...               ...               ...   
78612                 26           70                36          0.514286   
78613                 38          106                40          0.377358   
78614                 28          129                61          0.472868   
78615                 28          129                61          0.472868   
78616                 26           55                24          0.436364   

       punct_count_norm  norm_num_exclamations  \
0              0.153538               0.000000   
1              0.101010               0.000000   
2              0.095238               0.000000   
3              0.128686               0.000000   
4              0.153488               0.002326   
...                 ...                    ...   
78612          0.200000               0.000000   
78613          0.122642               0.000000   
78614          0.124031               0.000000   
78615          0.124031               0.000000   
78616          0.109091               0.000000   

                                        sentiment_scores  sentiment_scores_num  
0      {'neg': 0.048, 'neu': 0.866, 'pos': 0.086, 'co...                0.9847  
1      {'neg': 0.063, 'neu': 0.837, 'pos': 0.101, 'co...                0.9496  
2      {'neg': 0.049, 'neu': 0.907, 'pos': 0.044, 'co...               -0.6808  
3      {'neg': 0.066, 'neu': 0.875, 'pos': 0.059, 'co...               -0.2201  
4      {'neg': 0.043, 'neu': 0.898, 'pos': 0.059, 'co...                0.6683  
...                                                  ...                   ...  
78612  {'neg': 0.0, 'neu': 0.901, 'pos': 0.099, 'comp...                0.7650  
78613  {'neg': 0.082, 'neu': 0.891, 'pos': 0.026, 'co...               -0.6492  
78614  {'neg': 0.14

# Okay I neeed to clear my mind here. 


In [36]:
# separate data into true and false news
true_df = combined_df[combined_df['Label'] == True]
false_df = combined_df[combined_df['Label'] == False]

In [37]:
def t_test(true_data, fake_data):
    t_values = {}
    p_values = {}
    for col in true_data.columns:
        t, p = stats.ttest_ind(true_data[col], fake_data[col], equal_var=False, nan_policy='omit')
        t_values[col] = t
        p_values[col] = p
    return t_values, p_values


In [39]:
from scipy import stats
import numpy as np
# drop rows with missing or invalid values
combined_df = combined_df.dropna()
combined_df = combined_df.replace([np.inf, -np.inf], np.nan)
combined_df = combined_df.dropna()

# separate true and false dataframes
true_df = combined_df[combined_df['Label'] == True]
false_df = combined_df[combined_df['Label'] == False]

# perform t-test on numerical columns
for column in ['stop_words_count', 'word_count', 'cap_count', 'punct_count', 'num_exclamations', 'total_words', 'total_stop_words', 'stop_words_ratio', 'punct_count_norm', 'norm_num_exclamations', 'sentiment_scores', 'sentiment_scores_num']:
    t, p = stats.ttest_ind(true_df[column], false_df[column], equal_var=False, nan_policy='omit')
    print(column)
    print("T-test: t = %g  p-value = %g" % (t, p)) 


stop_words_count
T-test: t = 18.7199  p-value = 5.05826e-78
word_count
T-test: t = 22.9724  p-value = 2.15588e-116
cap_count
T-test: t = 0.843846  p-value = 0.398758
punct_count
T-test: t = 14.7224  p-value = 5.38372e-49
num_exclamations
T-test: t = -39.1612  p-value = 0
total_words
T-test: t = 22.6859  p-value = 1.44487e-113
total_stop_words
T-test: t = 18.1066  p-value = 3.99377e-73
stop_words_ratio
T-test: t = -38.7676  p-value = 0
punct_count_norm
T-test: t = -18.7714  p-value = 2.59673e-78
norm_num_exclamations
T-test: t = -27.6515  p-value = 6.43773e-167


TypeError: unsupported operand type(s) for +: 'dict' and 'dict'

# value insights

stop_words_count: The mean of stop words count is significantly higher for false news compared to true news.
word_count: The mean of word count is significantly higher for false news compared to true news.
punct_count: The mean of punctuation count is significantly higher for false news compared to true news.
num_exclamations: The mean of the number of exclamation marks is significantly higher for true news compared to false news.
total_words: The mean of total words is significantly higher for false news compared to true news.
total_stop_words: The mean of total stop words is significantly higher for false news compared to true news.
punct_count_norm: The mean of normalized punctuation count is significantly higher for true news compared to false news.
norm_num_exclamations: The mean of normalized number of exclamation marks is significantly higher for true news compared to false news.
These results suggest that false news articles tend to have more stop words, more words, more punctuation, and fewer total words than true news articles. On the other hand, true news articles tend to have more exclamation marks (both absolute and normalized) than false news articles.

In [40]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(combined_df['text'], combined_df['Label'], test_size=0.2, random_state=42)

# Create a TF-IDF Vectorizer object
tfidf = TfidfVectorizer(stop_words='english')

# Fit the Vectorizer on the training data
X_train_tfidf = tfidf.fit_transform(X_train)

# Train a Logistic Regression classifier on the TF-IDF features
clf = LogisticRegression()
clf.fit(X_train_tfidf, y_train)

# Transform the test data into TF-IDF features
X_test_tfidf = tfidf.transform(X_test)

# Evaluate the classifier on the test data
accuracy = clf.score(X_test_tfidf, y_test)
print(f"Accuracy: {accuracy}")

Accuracy: 0.9367209361485627


In [41]:
# Get the coefficients of the features
coefficients = clf.coef_[0]

# Get the indices that sort the coefficients
sorted_indices = coefficients.argsort()[::-1]

# Get the top 10 feature names
top_feature_names = [tfidf.get_feature_names_out()[i] for i in sorted_indices[:20]]

# Print the top 10 feature names
print("Top 20 features:")
for feature in top_feature_names:
    print(feature)



Top 20 features:
said
follow
thursday
tuesday
wednesday
monday
friday
twitter
breitbart
sunday
2017
reuters
saturday
ms
mr
edt
minister
reporters
statement
spokesman


In [42]:
combined_df

text  Label  \
0      The head of a conservative Republican faction ...   True   
1      Transgender people will be allowed for the fir...   True   
2      The special counsel investigation of links bet...   True   
3      Trump campaign adviser George Papadopoulos tol...   True   
4      President Donald Trump called on the U.S. Post...   True   
...                                                  ...    ...   
78612  The USA wants to divide Syria.\r\n\r\nGreat Br...  False   
78613  The Ukrainian coup d'etat cost the US nothing ...  False   
78614  The European Parliament falsifies history by d...  False   
78615  The European Parliament falsifies history by d...  False   
78616  A leading FSB officer, Segey Beseda, said duri...  False   

       stop_words_count  word_count  cap_count  punct_count  num_exclamations  \
0                   303         749        144          115                 0   
1                   171         396         72           40                 0   
2                   201         462        106           44                 0   
3                   153         373         99           48                 0   
4                   348         860        182          132                 2   
...                 ...         ...        ...          ...               ...   
78612                36          70         26           14                 0   
78613                40         106         38           13                 0   
78614                61         129         28           16                 0   
78615                61         129         28           16                 0   
78616                24          55         26            6                 0   

       cap_words_summary  total_words  total_stop_words  stop_words_ratio  \
0                    144          749               303          0.404539   
1                     72          396               171          0.431818   
2                    106          462               201          0.435065   
3                     99          373               153          0.410188   
4                    182          862               350          0.406032   
...                  ...          ...               ...               ...   
78612                 26           70                36          0.514286   
78613                 38          106                40          0.377358   
78614                 28          129                61          0.472868   
78615                 28          129                61          0.472868   
78616                 26           55                24          0.436364   

       punct_count_norm  norm_num_exclamations  \
0              0.153538               0.000000   
1              0.101010               0.000000   
2              0.095238               0.000000   
3              0.128686               0.000000   
4              0.153488               0.002326   
...                 ...                    ...   
78612          0.200000               0.000000   
78613          0.122642               0.000000   
78614          0.124031               0.000000   
78615          0.124031               0.000000   
78616          0.109091               0.000000   

                                        sentiment_scores  sentiment_scores_num  
0      {'neg': 0.048, 'neu': 0.866, 'pos': 0.086, 'co...                0.9847  
1      {'neg': 0.063, 'neu': 0.837, 'pos': 0.101, 'co...                0.9496  
2      {'neg': 0.049, 'neu': 0.907, 'pos': 0.044, 'co...               -0.6808  
3      {'neg': 0.066, 'neu': 0.875, 'pos': 0.059, 'co...               -0.2201  
4      {'neg': 0.043, 'neu': 0.898, 'pos': 0.059, 'co...                0.6683  
...                                                  ...                   ...  
78612  {'neg': 0.0, 'neu': 0.901, 'pos': 0.099, 'comp...                0.7650  
78613  {'neg': 0.082, 'neu': 0.891, 'pos': 0.026, 'co...               -0.6492  
78614  {'neg': 0.14

In [43]:
import matplotlib.pyplot as plt
import numpy as np

# Get the coefficients of the features
coefficients = clf.coef_[0]

# Get the indices that sort the coefficients
sorted_indices = coefficients.argsort()

# Get the top 10 feature names and coefficients
top_feature_names = [tfidf.get_feature_names_out()[i] for i in sorted_indices[-10:]]
top_coefficients = coefficients[sorted_indices][-10:]

# Create a horizontal bar chart
fig, ax = plt.subplots(figsize=(8, 6))
y_pos = np.arange(len(top_feature_names))
ax.barh(y_pos, top_coefficients, align='center')
ax.set_yticks(y_pos)
ax.set_yticklabels(top_feature_names)
ax.invert_yaxis()
ax.set_xlabel('Coefficient')
ax.set_title('Top 10 Indicative Features of True or False News')

# Save the plot as a PNG file
plt.savefig('top_features.png', dpi=300, bbox_inches='tight')


N-grams: You could extract n-grams (groups of n consecutive words) from the news body column and compare the frequency of different n-grams between true and false news.

Readability Metrics: Readability metrics are measures of how easy or difficult it is to read a piece of text. By calculating readability metrics such as Flesch-Kincaid Grade Level, Gunning Fog Index, or Coleman-Liau Index, you could compare the readability of true and false news. For example, if false news articles tend to have a lower readability score than true news articles, that could be an indicator of a trend.

Grammatical Errors: By analyzing the news body column for grammatical errors such as spelling mistakes, incorrect verb tense, or subject-verb agreement errors, you could compare the frequency of errors between true and false news. For example, if false news articles tend to have more grammatical errors than true news articles, that could be an indicator of a trend.

In [44]:
import pandas as pd
from nltk import ngrams
from collections import Counter


# Define the n-gram size
n = 3

# Define a function to extract n-grams from a string
def get_ngrams(text):
    # Tokenize the text into words
    words = text.split()
    # Extract n-grams
    ngrams_list = ngrams(words, n)
    # Convert the n-grams to strings
    ngrams_strings = [' '.join(ngram) for ngram in ngrams_list]
    return ngrams_strings

# Define a dictionary to hold the n-gram frequencies for true and false news
ngrams_freq = {True: Counter(), False: Counter()}

# Loop through each row in the DataFrame
for index, row in combined_df.iterrows():
    # Extract the news body and label from the row
    text = row['text']
    label = row['Label']
    # Extract the n-grams from the news body
    ngrams_list = get_ngrams(text)
    # Update the n-gram frequencies for the corresponding label
    ngrams_freq[label].update(ngrams_list)

# Print the most common n-grams for true and false news
print('Most common n-grams for true news:')
print(ngrams_freq[True].most_common(10))
print('Most common n-grams for false news:')
print(ngrams_freq[False].most_common(10))


Most common n-grams for true news:
[('the United States', 11412), ('said in a', 6921), ('one of the', 6495), ('the White House', 4727), ('President Donald Trump', 4230), ('as well as', 3751), ('in the United', 3636), ('a lot of', 3076), ('the end of', 3011), ('according to the', 2890)]
Most common n-grams for false news:
[('the United States', 6619), ('one of the', 6151), ('as well as', 3825), ('the fact that', 3793), ('a lot of', 3538), ('the White House', 3435), ('in order to', 3333), ('of the United', 3064), ('to be a', 2879), ('out of the', 2770)]


Both true and false news tend to have n-grams related to the United States, with "the United States" being the most common n-gram for both.
True news articles tend to have n-grams related to quotes and statements, with "said in a" and "according to the" being among the most common n-grams.
False news articles tend to have n-grams related to opinions and speculation, with "the fact that" being among the most common n-grams.
Both true and false news tend to have n-grams related to the White House, with "the White House" being among the most common n-grams for both.
Both true and false news articles tend to have n-grams related to government and politics, with "in the United" and "of the United" being among the most common n-grams for both.
False news articles tend to have more sensational or provocative n-grams, such as "to be a" and "out of the". This suggests that false news may be more likely to use exaggerated or attention-grabbing language.
True news articles tend to have more specific n-grams related to events or actions, such as "the end of" and "took place on". This suggests that true news may provide more detailed and factual information.
Both true and false news articles tend to use phrases like "as well as" and "a lot of", indicating that these are common filler phrases in news writing.

In [47]:
import pandas as pd
import textstat

# Define a function to calculate the readability metrics for a given text
def get_readability_metrics(text):
    fk_grade_level = textstat.flesch_kincaid_grade(text)
    gf_index = textstat.gunning_fog(text)
    return {'fk_grade_level': fk_grade_level, 'gf_index': gf_index}

# Create new columns for the readability metrics
combined_df['fk_grade_level'] = combined_df['text'].apply(textstat.flesch_kincaid_grade)
combined_df['gf_index'] = combined_df['text'].apply(textstat.gunning_fog)

# Calculate the average readability metrics for true and false news
true_fk_grade_level_avg = combined_df[combined_df['Label'] == True]['fk_grade_level'].mean()
false_fk_grade_level_avg = combined_df[combined_df['Label'] == False]['fk_grade_level'].mean()
true_gf_index_avg = combined_df[combined_df['Label'] == True]['gf_index'].mean()
false_gf_index_avg = combined_df[combined_df['Label'] == False]['gf_index'].mean()

# Print the results
print('Average Flesch-Kincaid Grade Level for true news: {:.2f}'.format(true_fk_grade_level_avg))
print('Average Flesch-Kincaid Grade Level for false news: {:.2f}'.format(false_fk_grade_level_avg))
print('Average Gunning Fog Index for true news: {:.2f}'.format(true_gf_index_avg))
print('Average Gunning Fog Index for false news: {:.2f}'.format(false_gf_index_avg))


Average Flesch-Kincaid Grade Level for true news: 10.91
Average Flesch-Kincaid Grade Level for false news: 10.35
Average Gunning Fog Index for true news: 12.42
Average Gunning Fog Index for false news: 11.83


Average Flesch-Kincaid Grade Level for true news: 10.91
Average Flesch-Kincaid Grade Level for false news: 10.35
Average Gunning Fog Index for true news: 12.42
Average Gunning Fog Index for false news: 11.83

Based on the average Flesch-Kincaid Grade Level and Gunning Fog Index values, it seems that there is not a significant difference in the readability of true and false news articles in the dataset. Both true and false news articles have similar scores, indicating that they may be similarly difficult to read.

In [48]:
from spellchecker import SpellChecker

# create a spellchecker object
spell = SpellChecker()

# define a function to count the number of misspelled words in a string
def count_misspelled_words(text):
    # split the text into words
    words = text.split()
    # count the number of misspelled words
    misspelled = spell.unknown(words)
    count = len(misspelled)
    return count

# apply the count_misspelled_words function to the news body column
combined_df['misspelled_words'] = combined_df['text'].apply(count_misspelled_words)


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.5/2.5 MB 3.3 MB/s eta 0:00:0000:0100:010m


In [49]:
# calculate the average number of misspelled words for true and false news articles
avg_misspelled_words_true = combined_df.loc[combined_df['Label'] == True, 'misspelled_words'].mean()
avg_misspelled_words_false = combined_df.loc[combined_df['Label'] == False, 'misspelled_words'].mean()

print('Average number of misspelled words for true news:', avg_misspelled_words_true)
print('Average number of misspelled words for false news:', avg_misspelled_words_false)


Average number of misspelled words for true news: 69.97415296640457
Average number of misspelled words for false news: 57.62813344942945


In [50]:
import spacy

# Load the English language model in spaCy
nlp = spacy.load('en_core_web_sm')

# Define a function to count the number of grammatical errors in a text
def count_grammar_errors(text):
    # Parse the text with spaCy
    doc = nlp(text)
    # Count the number of verb tense errors and subject-verb agreement errors
    num_errors = sum([token.tag_.startswith('V') and token.is_alpha and not token.is_stop and not token.is_oov and not token.is_punct and not token.is_digit and not token.like_num and (token.dep_ == 'aux' or token.dep_ == 'auxpass' or token.dep_ == 'VERB' and not token.head.tag_.startswith('V')) for token in doc])
    num_errors += sum([token.tag_ == 'VBZ' and token.is_alpha and not token.is_stop and not token.is_oov and not token.is_punct and not token.is_digit and not token.like_num and (token.dep_ == 'nsubj' or token.dep_ == 'nsubjpass') for token in doc])
    return num_errors

# Calculate the average number of grammar errors for true and false news
num_grammar_errors_true = [count_grammar_errors(text) for text in true_df['text']]
num_grammar_errors_false = [count_grammar_errors(text) for text in false_df['text']]
avg_num_grammar_errors_true = sum(num_grammar_errors_true) / len(num_grammar_errors_true)
avg_num_grammar_errors_false = sum(num_grammar_errors_false) / len(num_grammar_errors_false)

print(f'Average number of grammar errors for true news: {avg_num_grammar_errors_true:.2f}')
print(f'Average number of grammar errors for false news: {avg_num_grammar_errors_false:.2f}')


Average number of grammar errors for true news: 0.00
Average number of grammar errors for false news: 0.00


In [52]:
import spacy

# Load spaCy's English language model
nlp = spacy.load('en_core_web_sm')

# Define a function to extract named entities from text
def extract_entities(text):
    doc = nlp(text)
    entities = []
    for ent in doc.ents:
        entities.append((ent.text, ent.label_))
    return entities

# Apply the function to the 'text' column of the combined_df dataframe
combined_df['entities'] = combined_df['text'].apply(extract_entities)

# Print the top 10 most common named entities for true news
true_entities = combined_df[combined_df['Label'] == True]['entities'].sum()
true_entity_counts = Counter(true_entities)
print(true_entity_counts.most_common(10))

# Print the top 10 most common named entities for false news
false_entities = combined_df[combined_df['Label'] == False]['entities'].sum()
false_entity_counts = Counter(false_entities)
print(false_entity_counts.most_common(10))


[(('Trump', 'ORG'), 50667), (('U.S.', 'GPE'), 30168), (('Trump', 'PERSON'), 23460), (('Republican', 'NORP'), 21507), (('Clinton', 'PERSON'), 18792), (('first', 'ORDINAL'), 17332), (('the United States', 'GPE'), 16807), (('one', 'CARDINAL'), 16777), (('two', 'CARDINAL'), 16559), (('Donald Trump', 'PERSON'), 13515)]
[(('Trump', 'ORG'), 61723), (('Clinton', 'PERSON'), 21143), (('Donald Trump', 'PERSON'), 19017), (('US', 'GPE'), 17524), (('Russia', 'GPE'), 17316), (('one', 'CARDINAL'), 17047), (('American', 'NORP'), 15247), (('Hillary Clinton', 'PERSON'), 15103), (('America', 'GPE'), 14101), (('first', 'ORDINAL'), 13302)]
